src/main/java/com/atlassian/bitbucket/jenkins/internal/deployments/DeployedToEnvironmentNotifierStep.java

```diff
@@ -52,53 +49,41 @@
 
     private final String environmentKey;
```

Comment:
All of these can be `null` now. This is by design, as this is "end of the line" so to speak, they never leak out of this class (apart from rendering the UI, in which case `null` is fine) I felt this gave the absolutely best user experience

Reply:
I'm not certain we want to do this. `newInstance` means there is _always_ a check we can make before a stapler form is persisted, and although this constructor will probably only ever be used by the descriptor and tests, it is still an entity that should be able to be constructed as normal, and permitting nullable fields even with a javadoc comment feels a bit strange to me. If we can guarantee no nulls slip through stapler (and if they do, that's a bug in our code), why not permit null protection on fields explicitly expected to be non-null?